CANTroller2/src/display.h

```diff
@@ -559,22 +553,25 @@ class Display {
             if (halvings == 1) return ((color & 0xf000) | (color & 0x7c0) | (color & 0x1e)) >> 1;
             else return ((color & 0xe000) | (color & 0x780) | (color & 0x1c)) >> 2;
         }
-        void draw_idiotlight (int32_t index, int32_t x, int32_t y) {
-            _tft.fillRoundRect (x, y, 2 * disp_font_width + 1, disp_font_height + 1, 2, (*(idiotlights[index])) ? idiotcolors[index] : HGRY);  // GRY1);
-            _tft.setTextColor ((*(idiotlights[index])) ? BLK : darken_color (idiotcolors[index], 1));  // darken_color ((*(idiotlights[index])) ? BLK : DGRY)
-            _tft.setCursor (x+1, y+1);
-            _tft.print (idiotchars[index]);
-            idiotlasts[index] = *(idiotlights[index]);
-        }
-        void draw_idiotlights (int32_t x, int32_t y, bool force = false) {
-            for (int32_t ilite=0; ilite < arraysize(idiotlights); ilite++)
-                if (force || (*(idiotlights[ilite]) ^ idiotlasts[ilite]))
-                    draw_idiotlight (ilite, x + (2 * disp_font_width + 2) * ((ilite % disp_idiots_per_row) % disp_idiots_per_row), y + disp_idiot_row_height * (int32_t)(ilite / disp_idiots_per_row) + disp_idiot_row_height * (int32_t)(ilite / disp_idiots_per_row));
+        void draw_idiotlight(IdiotLight& light, int32_t x, int32_t y) {
+            _tft.fillRoundRect(x, y, 2 * disp_font_width + 1, disp_font_height + 1, 2, light.get_state() ? light.get_color() : darken_color(light.get_color(), 2));
+            _tft.setTextColor(light.get_state() ? BLK : GRY1);
+            _tft.setCursor(x+1, y+1);
+            _tft.print(light.get_label());
+            light.update();

```

Comment:
i thought we were updating them in a task? won't this slow down the main loop as well?

Reply:
Good catch. no need to update here if we're doing it in a task also